Select line plot units, sum spectra, export to ADS

jclarkeSTFC opened the pull request:
In the new Instrument View, add three (related) features:

- Change the unit of the line plot in the Instrument View using a drop-down
- Export the line plot to the ADS
- Sum (or not) the selected spectra

To sum spectra, the selected spectra are first converted to d-Spacing, rebinned, summed, then converted back to the desired unit if necessary. Once converted to d-Spacing, the selected spectra are rebinned to match the selected spectra with the widest x range. `SumSpectra` will not sum spectra with different bin edges.

Closes #38426 . <!-- One line per closed issue. -->

### To test:

Check all three new things together and separately. Right-click to select detectors.

<!-- Include sufficient instructions for someone unfamiliar with the application to test.
Ok to refer back to instructions in the issue.
-->

<!-- REMEMBER:
- Add labels, milestones, etc.
- Ensure the base of this PR is correct (e.g. release-next or main)
- Add release notes in separate file as per ([guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)), or justify their absence:
  *This does not require release notes* because <fill in an explanation of why>
-->

<!--  GATEKEEPER: When squashing, remove the section from HERE...  -->
---

### Reviewer

**Your comments will be used as part of the gatekeeper process.** Comment clearly on what you have checked and tested during your review. Provide an audit trail for any changes requested.

As per the [review guidelines](http://developer.mantidproject.org/ReviewingAPullRequest.html):

- Is the code of an acceptable quality? ([Code standards](http://developer.mantidproject.org/Standards/)/[GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html))
- Has a thorough functional test been performed? Do the changes handle unexpected input/situations?
- Are appropriately scoped unit and/or system tests provided?
- Do the release notes conform to the [guidelines](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html) and describe the changes appropriately?
- Has the relevant (user and developer) documentation been added/updated?

### Gatekeeper

As per the [gatekeeping guidelines](https://developer.mantidproject.org/Gatekeeping.html):

- Has a thorough first line review been conducted, including functional testing?
- At a high-level, is the code quality sufficient?
- Are the base, milestone and labels correct?
<!--  GATEKEEPER: ...To HERE  -->

jclarkeSTFC committed:
Allow selection of spectra line plot units

Spectra corresponding to picked detectors will now be extracted into
a separate workspace and converted to the selected units before
plotting.

jclarkeSTFC committed:
Add tests for unit selection.

jclarkeSTFC committed:
Add export to ADS and summing of spectra

Option to sum selected spectra, and a button to export the line plot
as a workspace to the ADS.

jclarkeSTFC committed:
Fixes following rebase

GuiMacielPereira reviewed (changes requested):
Everything seems to work well, but I noticed that the sum of spectra always gives the sum in TOF?
sum in TOF:
<img width="868" height="612" alt="image" src="https://github.com/user-attachments/assets/73d476b6-2628-461c-a40c-32c65ea6a452" />
And if I click to change to d-spacing:
<img width="868" height="612" alt="image" src="https://github.com/user-attachments/assets/3ea462f8-f49d-4828-a9f6-5d921191b914" />

jclarkeSTFC committed:
Rebin to ensure no loss of accuracy when summing

By taking the widest bin edge combined with the smallest bin width, we
ensure no loss of detail when summing the spectra together.

GuiMacielPereira reviewed (dismissed):
The sum is done in 'd-spacing' units and visually it looks accurate. The curve shape of the sum does not change between `TOF` and `d-spacing` units, but the axes are updated correctly.
Saving to the ADS is working and changing units changes the shape of the individual curves in the plot.

RichardWaiteSTFC reviewed (changes requested):
I have been playing around and have come across this error
```
Traceback (most recent call last):
  File "C:\Users\xhg73778\AppData\Local\mambaforge\envs\mantid-developer\Lib\site-packages\matplotlib\cbook.py", line 298, in process
    func(*args, **kwargs)
  File "C:\Users\xhg73778\AppData\Local\mambaforge\envs\mantid-developer\Lib\site-packages\matplotlib\backend_bases.py", line 2978, in mouse_move
    self.set_message(self._mouse_event_to_message(event))
  File "C:\mantid-conda\mantid\qt\python\mantidqt\mantidqt\plotting\mantid_navigation_toolbar.py", line 128, in set_message
    self.message.emit(s.replace("\n", ""))
    ^^^^^^^^^^^^
RuntimeError: wrapped C/C++ object of type WorkbenchNavigationToolbar has been deleted
```
Don't know how to reproduce it, but it might be something you can catch?

Also a few usability things:

1. I like there's a slider to control the height of the 1D plot - but on resizing/mouse release could you call `tight_layout` on the 1D plot so I can see axes and tick labels?
2. It's important to be able to zoom into the 1D plot to check exactly the d-spacing of the reflections. I can't seem to do this with the mouse, and double-clicking on the axes opens up a dialog box but it doesn't seem to update the limits (at least for me). Maybe it would be good to have some toolbar buttons as in normal matplotlib figures?
3. I'm used to the 1D plot axes limits (in TOF at least) being zoomed into the TOF limits selected in the 2D plot (in this case in the double slider) - this is minor more of a nice-to-have.
4. It would also be nice if you have zoomed into the 1D plot, if on changing x-units the axes limits are also converted? This may just be me though - feel free to push back or do in a separate PR!

RichardWaiteSTFC commented:
To be clear though - this is a lovely change and a big improvement - I really like being able to change the x-units of the 1D plot!

jclarkeSTFC committed:
Merge branch 'main' into select_line_plot_units

jclarkeSTFC committed:
Add momentum transfer to line plot units

jclarkeSTFC committed:
When moving splitters, call tight_layout on the plot

This ensures the plot axes are displayed correctly on resizing

jclarkeSTFC committed:
Acccount for ragged workspaces when rebinning

Calling extractX on a ragged workspace will truncate the bin edges to
the number of bin edges of the spectra with the fewest edges. So we
need to loop over each spectra individually to get the rebin parameters.

Also, no need to convert units if the workspace is already in d-Spacing.

jclarkeSTFC committed:
Handle workspaces with no units on the x-axis

If the workspace has no units on the x-axis, then we can't convert to
any other unit. We can still rebin, sum spectra, and export.

jclarkeSTFC committed:
Set the line plot to the workspace unit

If available, set the unit of the line plot equal to the workspace unit
on opening.

RichardWaiteSTFC commented on qt/python/instrumentview/instrumentview/FullInstrumentViewModel.py:
```diff
@@ -200,15 +200,24 @@ def extract_spectra_for_line_plot(self, unit: str, sum_spectra: bool) -> None:
         if sum_spectra and len(workspace_indices) > 1:
             # Sum in d-Spacing to avoid blurring peaks
-            ws = ConvertUnits(InputWorkspace=ws, target="dSpacing", EMode="Elastic", EnableLogging=False, StoreInADS=False)
+            if ws.getAxis(0).getUnit().unitID() != "dSpacing":
```
Ah sorry - I'm going to request that actually we don't convert units here (at least not to the whole workspace - which could be wasteful anyway).
I think it's my fault for not being clear in previous discussions (here and offline)!

Here is the behaviour I think we should aim for:

- ws and 1D plot unit are the same (e.g. TOF) and `ws.isCommonBins=True`: `SumSpectra` (no need to extract or rebin)
- ws and 1D plot unit are the same and  `ws.isCommonBins=False` (e.g. d-spacing): `ExtractSpectra` >  `Rebin` > `SumSpectra`
- ws and 1D plot unit are the different: `ExtractSpectra` > `ConvertUnits` > if `ws_extracted.isCommonBins()=False` then `Rebin` > `SumSpectra`

Do you agree? Have I missed any scenario?

jclarkeSTFC replied:
I've changed it so it will sum in the selected unit, instead of always in d-spacing.

RichardWaiteSTFC reviewed (changes requested):
Thanks for the changes, all working well - just a clarification about expected behaviour of summing spectra (sorry I think I haven't been as clear about this as I should have).

Also it would be nice to have a mode where when you click the 1D plot is replaced (i.e. multiple spectra are not summed, or overplotted).

jclarkeSTFC committed:
Sum selected spectra in selected unit

RichardWaiteSTFC reviewed (approved):
Looks great thanks for this!
